js/dispatch.ts

```diff
@@ -1,105 +1,233 @@
-export const OP_COPY_FILE = 47;
-export const OP_STAT = 48;
-export const OP_READ_DIR = 49;
-export const OP_RENAME = 50;
-export const OP_LINK = 51;
-export const OP_SYMLINK = 52;
-export const OP_READ_LINK = 53;
-export const OP_TRUNCATE = 54;
-export const OP_MAKE_TEMP_DIR = 55;
-export const OP_CWD = 56;
-export const OP_FETCH_ASSET = 57;
+const opNamespace = ops.builtins;
+
+type MaybeOpId = number | undefined;
 
-export function asyncMsgFromRust(opId: number, ui8: Uint8Array): void {
-  switch (opId) {
-    case OP_WRITE:
-    case OP_READ:
-      minimal.asyncMsgFromRust(opId, ui8);
-      break;
-    case OP_EXIT:
-    case OP_IS_TTY:
-    case OP_ENV:
-    case OP_EXEC_PATH:
-    case OP_UTIME:
-    case OP_OPEN:
-    case OP_SEEK:
-    case OP_FETCH:
-    case OP_REPL_START:
-    case OP_REPL_READLINE:
-    case OP_ACCEPT:
-    case OP_DIAL:
-    case OP_GLOBAL_TIMER:
-    case OP_HOST_GET_WORKER_CLOSED:
-    case OP_HOST_GET_MESSAGE:
-    case OP_WORKER_GET_MESSAGE:
-    case OP_RUN_STATUS:
-    case OP_MKDIR:
-    case OP_CHMOD:
-    case OP_CHOWN:
-    case OP_REMOVE:
-    case OP_COPY_FILE:
-    case OP_STAT:
-    case OP_READ_DIR:
-    case OP_RENAME:
-    case OP_LINK:
-    case OP_SYMLINK:
-    case OP_READ_LINK:
-    case OP_TRUNCATE:
-    case OP_MAKE_TEMP_DIR:
-      json.asyncMsgFromRust(opId, ui8);
-      break;
-    default:
-      throw Error("bad async opId");
-  }
-}
+type OpId = number;
+
+// These consts are shared with Rust. Update with care.
+export let OP_READ: OpId;
+opNamespace.read = (id: MaybeOpId): void => {
+  OP_READ = id!;
+  core.setAsyncHandler(
+    id!,
+    (buf: Uint8Array): void => minimal.asyncMsgFromRust(id!, buf)
+  );
+};
+export let OP_WRITE: OpId;
+opNamespace.write = (id: MaybeOpId): void => {
+  OP_WRITE = id!;
+  core.setAsyncHandler(
+    id!,
+    (buf: Uint8Array): void => minimal.asyncMsgFromRust(id!, buf)
+  );
+};
+export let OP_EXIT: OpId;
+opNamespace.exit = (id: MaybeOpId): void => {
+  OP_EXIT = id!;
+  core.setAsyncHandler(id!, json.asyncMsgFromRust);
+};
+export let OP_IS_TTY: OpId;
+opNamespace.isTty = (id: MaybeOpId): void => {
+  OP_IS_TTY = id!;
+  core.setAsyncHandler(id!, json.asyncMsgFromRust);
+};
+export let OP_ENV: OpId;
+opNamespace.env = (id: MaybeOpId): void => {
+  OP_ENV = id!;
+  core.setAsyncHandler(id!, json.asyncMsgFromRust);
+};
+export let OP_EXEC_PATH: OpId;
+opNamespace.execPath = (id: MaybeOpId): void => {
+  OP_EXEC_PATH = id!;
+  core.setAsyncHandler(id!, json.asyncMsgFromRust);
+};
+export let OP_SET_ENV: OpId;
+opNamespace.setEnv = (id: MaybeOpId): void => {
+  OP_SET_ENV = id!;
+  core.setAsyncHandler(id!, json.asyncMsgFromRust);
+};
+export let OP_HOME_DIR: OpId;
+opNamespace.homeDir = (id: MaybeOpId): void => {
+  OP_HOME_DIR = id!;
+  core.setAsyncHandler(id!, json.asyncMsgFromRust);
+};
+export let OP_START: OpId;
+opNamespace.start = (id: MaybeOpId): void => {
+  OP_START = id!;
+  core.setAsyncHandler(id!, json.asyncMsgFromRust);
```

Comment:
This API is strange...

Reply:
I'm using a class based API like the one you suggested in all new implementations. Consider this depreciated.